distributed/tests/test_scheduler.py

```diff
@@ -4099,3 +4100,112 @@ async def test_count_task_prefix(c, s, a, b):
 
     assert s.task_prefixes["inc"].state_counts["memory"] == 20
     assert s.task_prefixes["inc"].state_counts["erred"] == 0
+
+
+@gen_cluster(client=True)
+async def test_transition_waiting_memory(c, s, a, b):
+    """Test race condition where a task transitions to memory while its state on the
+    scheduler is waiting:
+
+    1. worker a finishes x
+    2. y transitions to processing and is assigned to worker b
+    3. b fetches x and sends an add_keys message to the scheduler
+    4. In the meantime, a dies and causes x to be scheduled back to released/waiting.
+    5. Scheduler queues up a free-keys intended for b to cancel both x and y
+    6. Before free-keys arrives to b, the worker runs and completes y, sending a
+       finished-task message to the scheduler
+    7. {op: add-keys, keys=[x]} from b finally arrives to the scheduler. This triggers
+       a {op: remove-replicas, keys=[x]} message from the scheduler to worker b, because
+       add-keys when the task state is not memory triggers a cleanup of redundant
+       replicas (see Scheduler.add_keys) - in this, add-keys differs from finished-task!
+    8. {op: task-finished, key=y} from b arrives to the scheduler
+
+    See also:
+    - test_transition_no_worker_memory
+    - test_transition_queued_memory
+    - test_transition_processing_memory_from_unexpected_worker
+    """
+    # This test is very timing-sensitive. Instead of using inc, which would need to
```

Comment:
I wonder if you could make this less timing-sensitive by using `freeze_batched_send`?

https://github.com/dask/distributed/blob/abf36de1a6d075fd49f0ee96ccde85a581a27ea9/distributed/utils_test.py#L2325-L2352

Reply:
Done - it's much cleaner now, ty!